sotodlib/hwp/g3thwp.py

```diff
@@ -905,87 +958,35 @@ def write_solution_h5(self, tod, output=None, h5_address=None, load_h5=True):
-
-            - raw_quad_1/2: bool (2, x)
-
-            - raw_pid_direction: bool (2, x)
 
         """
-        if self._output is None and output is None:
-            logger.warning('Output file not specified')
-            return
-        if output is not None:
-            self._output = output
-
-        if len(tod.timestamps) == 0:
-            logger.warning('Input data is empty.')
-            return
 
         # write solution, metadata loader requires a dets axis
         aman = sotodlib.core.AxisManager(tod.dets, tod.samps)
         aman.wrap_new('timestamps', ('samps', ))[:] = tod.timestamps
         self._set_empty_axes(aman)
 
-        start = int(tod.timestamps[0])-self._margin
-        end = int(tod.timestamps[-1])+self._margin
-
-        data = {}
-        try:
-            if load_h5:
-                logger.info('Loading raw encoder data from h5')
-                try:
-                    data = self._load_raw_axes(aman, output, h5_address)
-                except Exception as e:
-                    logger.error(f"Exception '{e}' thrown while loading HWP data from h5. Attempt to load from hk.")
-                    data = self.load_data(start, end)
-
+        if 'pid_direction' in self.data.keys():
+            pid_direction = np.nanmedian(self.data['pid_direction'][1])*2 - 1
+            if pid_direction in [1, -1]:
+                aman['pid_direction'] = pid_direction
             else:
-                data = self.load_data(start, end)
-
-            if 'pid_direction' in data.keys():
-                pid_direction = np.nanmedian(data['pid_direction'][1])*2 - 1
-                if pid_direction in [1, -1]:
-                    aman['pid_direction'] = pid_direction
-                else:
-                    aman['pid_direction'] = 0
-
-            logger.info('Saving raw encoder data')
-            self._set_raw_axes(aman, data)
-
-        except Exception as e:
-            logger.error(
-                f"Exception '{e}' thrown while loading HWP data. The specified encoder field is missing.")
-            self._write_solution_h5_logger = 'HWP data too short'
-            print(traceback.format_exc())
+                aman['pid_direction'] = 0
 
         solved = {}
         for suffix in self._suffixes:
             logger.info('Start analyzing encoder'+suffix)
             self._set_empty_axes(aman, suffix)
             # load data
-            if not 'counter' + suffix in data.keys():
+            if not 'counter' + suffix in self.data.keys():
                 logger.warning('No HWP data in the specified timestamps.')
                 self._write_solution_h5_logger = 'No HWP data'
                 continue
 
             # version 1
             # calculate HWP angle
             try:
-                solved = solved | self.analyze(data, mod2pi=False, suffix=suffix)
+                solved = solved | self.analyze(self.data, mod2pi=False, suffix=suffix)
```

Comment:
This is a very strange line... the operator `|` isn't defined for dicts.  Am I missing something?

Reply:
You probably mean "or", right?  Has this code been tested?

Reply:
This is merging two dictionaries for two encoders. This way is supported from python 3.9.0
https://stackoverflow.com/questions/38987/how-do-i-merge-two-dictionaries-in-a-single-expression-in-python

Reply:
And yes, this is tested. We have minor modification in this PR, but the this code (operator `|`) has been used from previous version of metadata 240301m.

Reply:
Ah, I have not seen that... and was checking on python 3.8.

We are allowed to require py3.9 now but I will note you could have simply done
```
solved.update(self.analyze(self.data, mod2pi=False, suffix=suffix))
```
... and that would have been python2 compatible!

Thanks.